README.md
# eBay Provider for OAuth 2.0 Client

This package provides an eBay OAuth 2.0 support for the PHP League's [OAuth 2.0 Client](https://github.com/thephpleague/oauth2-client).

## Installation

To install, use composer:

```
composer require nitromike502/ebay-api-oauth2
```

## Usage

Usage is the same as The League's OAuth client, using `\Nitromike502\OAuth2\Client\Provider\Ebay` as the provider.

### Authorization Code Flow

```php
$provider = new Nitromike502\OAuth2\Client\Provider\Ebay([
    'clientId'          => '{ebay-client-id}',
    'clientSecret'      => '{ebay-client-secret}',
    'redirectUri'       => '{redirect-code}'
]);
```
For further usage of this package please refer to the [core package documentation on "Authorization Code Grant"](https://github.com/thephpleague/oauth2-client#usage).

### Refreshing a Token

```php
$provider = new Nitromike502\OAuth2\Client\Provider\Ebay([
    'clientId'          => '{ebay-client-id}',
    'clientSecret'      => '{ebay-client-secret}',
    'redirectUri'       => '{redirect-code}'
]);

$existingAccessToken = getAccessTokenFromYourDataStore();

if ($existingAccessToken->hasExpired()) {
    $newAccessToken = $provider->getAccessToken('refresh_token', [
        'refresh_token' => $existingAccessToken->getRefreshToken()
    ]);

    // Purge old access token and store new access token to your data store.
}
```

For further usage of this package please refer to the [core package documentation on "Refreshing a Token"](https://github.com/thephpleague/oauth2-client#refreshing-a-token).

## Testing

``` bash
$ ./vendor/bin/phpunit
```

## License

The MIT License (MIT). Please see [License File](https://github.com/nitromike502/ebay-api-oauth2/blob/master/LICENSE) for more information.


CHANGELOG.md
## 0.1.0

_Released: 2020-08-19

* Initial release.


tests/src/EbayTest.php
<?php
namespace Nitromike502\OAuth2\Client\Test;

use GuzzleHttp\ClientInterface;
use Nitromike502\OAuth2\Client\Provider\Ebay;
use League\OAuth2\Client\Token\AccessToken;
use PHPUnit\Framework\TestCase;
use Psr\Http\Message\ResponseInterface;

class EbayTest extends TestCase
{
    /**
     * @var Ebay
     */
    protected $provider;

    protected function setUp()
    {
        $this->provider = new Ebay([
            'clientId' => 'mock_client_id',
            'clientSecret' => 'mock_secret',
            'redirectUri' => 'mock_redirect_uri'
        ]);
    }

    protected function getJsonFile($file, $encode = false)
    {
        $json = file_get_contents(__DIR__ . '/../' . $file);
        $data = json_decode($json, true);

        if ($encode && json_last_error() == JSON_ERROR_NONE) {
            return $data;
        }

        return $json;
    }

    public function testAuthorizationUrl()
    {
        $url = $this->provider->getAuthorizationUrl();
        $uri = parse_url($url);
        parse_str($uri['query'], $query);

        $this->assertArrayHasKey('client_id', $query);
        $this->assertArrayHasKey('redirect_uri', $query);
        $this->assertArrayHasKey('state', $query);
        $this->assertArrayHasKey('scope', $query);
        $this->assertArrayHasKey('response_type', $query);
        $this->assertNotNull($this->provider->getState());
    }

    public function testScopes()
    {
        $options = [
            'scope' => [uniqid(), uniqid()]
        ];

        $url = $this->provider->getAuthorizationUrl($options);

        $this->assertContains(implode('%20', $options['scope']), $url);
    }

    public function testGetAuthorizationUrl()
    {
        $this->provider->enableSandbox(false);
        $url = $this->provider->getAuthorizationUrl();
        $uri = parse_url($url);

        $this->assertEquals('api.ebay.com', $uri['host']);
        $this->assertEquals('/oauth2/authorize', $uri['path']);
        
        $this->provider->enableSandbox(true);
        $url = $this->provider->getAuthorizationUrl();
        $uri = parse_url($url);

        $this->assertEquals('api.sandbox.ebay.com', $uri['host']);
        $this->assertEquals('/oauth2/authorize', $uri['path']);
    }

    public function testGetBaseAccessTokenUrl()
    {
        $params = [];

        $this->provider->enableSandbox(false);
        $url = $this->provider->getBaseAccessTokenUrl($params);
        $uri = parse_url($url);

        $this->assertEquals('api.ebay.com', $uri['host']);
        $this->assertEquals('/identity/'.$this->provider->getApiVersion().'/oauth2/token', $uri['path']);
        
        $this->provider->enableSandbox(true);
        $url = $this->provider->getBaseAccessTokenUrl($params);
        $uri = parse_url($url);

        $this->assertEquals('api.sandbox.ebay.com', $uri['host']);
        $this->assertEquals('/identity/'.$this->provider->getApiVersion().'/oauth2/token', $uri['path']);
    }

    public function testGetResourceOwnerDetailsUrl()
    {
        $token = 'mock_token';
        $tokenMock = $this->getMockBuilder(AccessToken::class)
            ->disableOriginalConstructor()
            ->getMock();

        $tokenMock->method('getToken')->willReturn($token);

        $this->provider->enableSandbox(false);
        $url = $this->provider->getResourceOwnerDetailsUrl($tokenMock);
        $uri = parse_url($url);

        $this->assertEquals('apiz.ebay.com', $uri['host']);
        $this->assertEquals('/commerce/identity/v1/user/', $uri['path']);
        
        $this->provider->enableSandbox(true);
        $url = $this->provider->getResourceOwnerDetailsUrl($tokenMock);
        $uri = parse_url($url);

        $this->assertEquals('apiz.sandbox.ebay.com', $uri['host']);
        $this->assertEquals('/commerce/identity/v1/user/', $uri['path']);
    }

    public function testGetAccessToken()
    {
        $accessToken = $this->getJsonFile('access_token_response.json');
        $responseMock = $this->getMockBuilder(ResponseInterface::class)->getMock();
        $responseMock->method('getBody')->willReturn($accessToken);
        $responseMock->method('getHeader')->willReturn(['content-type' => 'json']);
        $responseMock->method('getStatusCode')->willReturn(200);

        $clientMock = $this->getMockBuilder(ClientInterface::class)->getMock();
        $clientMock->expects($this->once())->method('send')->willReturn($responseMock);
        $this->provider->setHttpClient($clientMock);

        $token = $this->provider->getAccessToken('authorization_code', ['code' => 'mock_authorization_code']);

        $this->assertEquals('mock_access_token', $token->getToken());
        $this->assertEquals(time() + 7200, $token->getExpires());
        $this->assertEquals('mock_refresh_token', $token->getRefreshToken());
        $this->assertNull($token->getResourceOwnerId());
    }

    public function testUserData()
    {
        $accessToken = $this->getJsonFile('access_token_response.json');
        $accessTokenInfoJson = $this->getJsonFile('access_token_info.json');
        $accessTokenInfo = json_decode($accessTokenInfoJson, true);

        $postResponseMock = $this->getMockBuilder(ResponseInterface::class)->getMock();
        $postResponseMock->method('getBody')->willReturn($accessToken);
        $postResponseMock->method('getHeader')->willReturn(['content-type' => 'json']);
        $postResponseMock->method('getStatusCode')->willReturn(200);

        $accessTokenInfoResponseMock = $this->getMockBuilder(ResponseInterface::class)->getMock();
        $accessTokenInfoResponseMock->method('getBody')->willReturn($accessTokenInfoJson);
        $accessTokenInfoResponseMock->method('getHeader')->willReturn(['content-type' => 'json']);
        $accessTokenInfoResponseMock->method('getStatusCode')->willReturn(200);

        $clientMock = $this->getMockBuilder(ClientInterface::class)->getMock();
        $clientMock->expects($this->exactly(2))
            ->method('send')
            ->willReturnOnConsecutiveCalls($postResponseMock, $accessTokenInfoResponseMock);

        $this->provider->setHttpClient($clientMock);

        $token = $this->provider->getAccessToken('authorization_code', ['code' => 'mock_authorization_code']);
        $user = $this->provider->getResourceOwner($token);

        $this->assertEquals($accessTokenInfo, $user->toArray());
        $this->assertEquals($accessTokenInfo['userId'], $user->getUserId());
        $this->assertEquals($accessTokenInfo['userId'], $user->getId());
        $this->assertEquals($accessTokenInfo['username'], $user->getUsername());
        $this->assertEquals($accessTokenInfo['accountType'], $user->getAccountType());
        $this->assertEquals($accessTokenInfo['registrationMarketplaceId'], $user->getRegistrationMarketplaceId());
    }

    /**
     * @expectedException \League\OAuth2\Client\Provider\Exception\IdentityProviderException
     */
    public function testExceptionThrownWhenErrorReceived()
    {
        $status = rand(401,599);
        $postResponseMock = $this->getMockBuilder(ResponseInterface::class)->getMock();
        $postResponseMock->method('getBody')->willReturn('{"error": "error_code","error_description": "A human readable error message"}');
        $postResponseMock->method('getHeader')->willReturn(['content-type' => 'json']);
        $postResponseMock->method('getStatusCode')->willReturn($status);

        $clientMock = $this->getMockBuilder(ClientInterface::class)->getMock();
        $clientMock->expects($this->once())->method('send')->willReturn($postResponseMock);

        $this->provider->setHttpClient($clientMock);

        $this->provider->getAccessToken('authorization_code', ['code' => 'mock_authorization_code']);
    }
}


src/Provider/Ebay.php
<?php

namespace Nitromike502\OAuth2\Client\Provider;

use League\OAuth2\Client\Token\AccessToken;
use League\OAuth2\Client\Provider\AbstractProvider;
use Nitromike502\OAuth2\Client\Provider\Exception\EbayProviderException;
use Psr\Http\Message\ResponseInterface;
use League\OAuth2\Client\Tool\BearerAuthorizationTrait;

class Ebay extends AbstractProvider
{
    use BearerAuthorizationTrait;
    
    const ACCESS_TOKEN_RESOURCE_OWNER_ID = 'individualAccount';
    
    const BASE_SANDBOX_URL = 'https://api.sandbox.ebay.com';
    const BASE_PRODUCTION_URL = 'https://api.ebay.com';
    const AUTHORIZE_SANDBOX_URL = 'https://auth.sandbox.ebay.com';
    const AUTHORIZE_PRODUCTION_URL = 'https://auth.ebay.com';
    
    private $_enableSandboxMode;

    private $api_version = 'v1';
       
    protected $scopes = [];
    
    /**
     * @param array $options
     * @param array $collaborators
     *
     * @throws \InvalidArgumentException
     */
    public function __construct($options = [], array $collaborators = [])
    {
        if (array_key_exists('sandbox',$options))
        {
            $this->enableSandbox((bool) $options['sandbox']);
        }
        $this->clientId = $options['clientId'];
        $this->clientSecret = $options['clientSecret'];
        $this->redirectUri = $options['redirectUri'];
        
        if (isset($options['scopes']))
        {
            $this->setScopes($options['scopes']);
        }
        
        parent::__construct($options, $collaborators);
    }
    
    public function enableSandbox($enable = true)
    {
        $this->_enableSandboxMode = $enable;
    }
    
    public function getBaseEbayUrl()
    {
        return $this->_enableSandboxMode ? static::BASE_SANDBOX_URL : static::BASE_PRODUCTION_URL;
    }
    
    public function setApiVersion($version)
    {
        $this->api_version = $version;
        return $this;
    }
    
    public function getApiVersion()
    {
        return $this->api_version;
    }
    
    /**
     * params for application user authorization request
     */
    protected function getAuthorizationParameters(array $options)
    {
        $params = parent::getAuthorizationParameters($options);
        $params['redirect_uri'] = $this->redirectUri;
        unset($params['approval_prompt']);
        return $params;
    }

    /**
     * Generates URL for user to authorize aplication
     */
    public function getBaseAuthorizationUrl()
    {
        $url = $this->_enableSandboxMode ? static::AUTHORIZE_SANDBOX_URL : static::AUTHORIZE_PRODUCTION_URL;
        return $url.'/oauth2/authorize/';
    }

    /**
     * generates base acccess token URL
     */
    public function getBaseAccessTokenUrl(array $params)
    {
        return $this->getBaseEbayUrl().'/identity/'.$this->api_version.'/oauth2/token';
    }

    public function setScopes($scopes = [])
    {
        $scopes = array_unique(array_filter($this->scopes));
        $scopes[] = '';
        $this->scopes = $scopes;
        return $this;
    }
    
    public function getDefaultScopes()
    {
        return $this->scopes;
    }

    public function getResourceOwnerDetailsUrl(AccessToken $token)
    {
        return $this->_enableSandboxMode 
            ? 'https://apiz.sandbox.ebay.com/commerce/identity/v1/user/' 
            : 'https://apiz.ebay.com/commerce/identity/v1/user/';
    }

    protected function createResourceOwner(array $response, AccessToken $token)
    {
        return new EbayUser($response);
    }

    protected function checkResponse(ResponseInterface $response, $data)
    {
        if (!empty($data['error'])) 
        {
            $message = $data['error_description'];
            throw new EbayProviderException($message, 0, $data);
        }
    }
    
    /**
     * Returns the string that should be used to separate scopes when building
     * the URL for requesting an access token.
     *
     * @return string Scope separator, defaults to ','
     */
    protected function getScopeSeparator()
    {
        return ' ';
    }
}












src - Copy/Provider/EbayUser.php
<?php

namespace Nitromike502\OAuth2\Client\Provider;

use League\OAuth2\Client\Provider\ResourceOwnerInterface;

class EbayUser implements ResourceOwnerInterface
{
    /**
     * @var array
     */
    protected $data;

    /**
     * @param  array $response
     */
    public function __construct(array $response)
    {
        $this->data = $response;
    }

	public function getId()
	{
		return $this->getUserId();
	}
	
    /**
     * Returns the ID for the user as a string if present.
     *
     * @return string|null
     */
    public function getUserId()
    {
        return $this->getField('userId');
    }

    /**
     * Returns the username for the user as a string if present.
     *
     * @return string|null
     */
    public function getUsername()
    {
        return $this->getField('username');
    }

    /**
     * Returns the account type for the user as a string if present.
     *
     * @return string|null
     */
    public function getAccountType()
    {
        return $this->getField('accountType');
    }

    /**
     * Returns the last name for the user as a string if present.
     *
     * @return string|null
     */
    public function getRegistrationMarketplaceId()
    {
        return $this->getField('registrationMarketplaceId');
    }

    /**
     * Returns all the data obtained about the user.
     *
     * @return array
     */
    public function toArray()
    {
        return $this->data;
    }

    /**
     * Returns a field from the Graph node data.
     *
     * @param string $key
     *
     * @return mixed|null
     */
    private function getField($key)
    {
        return isset($this->data[$key]) ? $this->data[$key] : null;
    }
}

tests - Copy/tests - Copy - Copy/src/EbayTest.php
<?php
namespace Nitromike502\OAuth2\Client\Test;

use GuzzleHttp\ClientInterface;
use Nitromike502\OAuth2\Client\Provider\Ebay;
use League\OAuth2\Client\Token\AccessToken;
use PHPUnit\Framework\TestCase;
use Psr\Http\Message\ResponseInterface;

class EbayTest extends TestCase
{
    /**
     * @var Ebay
     */
    protected $provider;

    protected function setUp()
    {
        $this->provider = new Ebay([
            'clientId' => 'mock_client_id',
            'clientSecret' => 'mock_secret',
            'redirectUri' => 'mock_redirect_uri'
        ]);
    }

    protected function getJsonFile($file, $encode = false)
    {
        $json = file_get_contents(__DIR__ . '/../' . $file);
        $data = json_decode($json, true);

        if ($encode && json_last_error() == JSON_ERROR_NONE) {
            return $data;
        }

        return $json;
    }

    public function testAuthorizationUrl()
    {
        $url = $this->provider->getAuthorizationUrl();
        $uri = parse_url($url);
        parse_str($uri['query'], $query);

        $this->assertArrayHasKey('client_id', $query);
        $this->assertArrayHasKey('redirect_uri', $query);
        $this->assertArrayHasKey('state', $query);
        $this->assertArrayHasKey('scope', $query);
        $this->assertArrayHasKey('response_type', $query);
        $this->assertNotNull($this->provider->getState());
    }

    public function testScopes()
    {
        $options = [
            'scope' => [uniqid(), uniqid()]
        ];

        $url = $this->provider->getAuthorizationUrl($options);

        $this->assertContains(implode('%20', $options['scope']), $url);
    }

    public function testGetAuthorizationUrl()
    {
        $this->provider->enableSandbox(false);
        $url = $this->provider->getAuthorizationUrl();
        $uri = parse_url($url);

        $this->assertEquals('auth.ebay.com', $uri['host']);
        $this->assertEquals('/oauth2/authorize/', $uri['path']);
        
        $this->provider->enableSandbox(true);
        $url = $this->provider->getAuthorizationUrl();
        $uri = parse_url($url);

        $this->assertEquals('auth.sandbox.ebay.com', $uri['host']);
        $this->assertEquals('/oauth2/authorize/', $uri['path']);
    }

    public function testGetBaseAccessTokenUrl()
    {
        $params = [];

        $this->provider->enableSandbox(false);
        $url = $this->provider->getBaseAccessTokenUrl($params);
        $uri = parse_url($url);

        $this->assertEquals('api.ebay.com', $uri['host']);
        $this->assertEquals('/identity/'.$this->provider->getApiVersion().'/oauth2/token', $uri['path']);
        
        $this->provider->enableSandbox(true);
        $url = $this->provider->getBaseAccessTokenUrl($params);
        $uri = parse_url($url);

        $this->assertEquals('api.sandbox.ebay.com', $uri['host']);
        $this->assertEquals('/identity/'.$this->provider->getApiVersion().'/oauth2/token', $uri['path']);
    }

    public function testGetResourceOwnerDetailsUrl()
    {
        $token = 'mock_token';
        $tokenMock = $this->getMockBuilder(AccessToken::class)
            ->disableOriginalConstructor()
            ->getMock();

        $tokenMock->method('getToken')->willReturn($token);

        $this->provider->enableSandbox(false);
        $url = $this->provider->getResourceOwnerDetailsUrl($tokenMock);
        $uri = parse_url($url);

        $this->assertEquals('apiz.ebay.com', $uri['host']);
        $this->assertEquals('/commerce/identity/v1/user/', $uri['path']);
        
        $this->provider->enableSandbox(true);
        $url = $this->provider->getResourceOwnerDetailsUrl($tokenMock);
        $uri = parse_url($url);

        $this->assertEquals('apiz.sandbox.ebay.com', $uri['host']);
        $this->assertEquals('/commerce/identity/v1/user/', $uri['path']);
    }

    public function testGetAccessToken()
    {
        $accessToken = $this->getJsonFile('access_token_response.json');
        $responseMock = $this->getMockBuilder(ResponseInterface::class)->getMock();
        $responseMock->method('getBody')->willReturn($accessToken);
        $responseMock->method('getHeader')->willReturn(['content-type' => 'json']);
        $responseMock->method('getStatusCode')->willReturn(200);

        $clientMock = $this->getMockBuilder(ClientInterface::class)->getMock();
        $clientMock->expects($this->once())->method('send')->willReturn($responseMock);
        $this->provider->setHttpClient($clientMock);

        $token = $this->provider->getAccessToken('authorization_code', ['code' => 'mock_authorization_code']);

        $this->assertEquals('mock_access_token', $token->getToken());
        $this->assertEquals(time() + 7200, $token->getExpires());
        $this->assertEquals('mock_refresh_token', $token->getRefreshToken());
        $this->assertNull($token->getResourceOwnerId());
    }

    public function testUserData()
    {
        $accessToken = $this->getJsonFile('access_token_response.json');
        $accessTokenInfoJson = $this->getJsonFile('access_token_info.json');
        $accessTokenInfo = json_decode($accessTokenInfoJson, true);

        $postResponseMock = $this->getMockBuilder(ResponseInterface::class)->getMock();
        $postResponseMock->method('getBody')->willReturn($accessToken);
        $postResponseMock->method('getHeader')->willReturn(['content-type' => 'json']);
        $postResponseMock->method('getStatusCode')->willReturn(200);

        $accessTokenInfoResponseMock = $this->getMockBuilder(ResponseInterface::class)->getMock();
        $accessTokenInfoResponseMock->method('getBody')->willReturn($accessTokenInfoJson);
        $accessTokenInfoResponseMock->method('getHeader')->willReturn(['content-type' => 'json']);
        $accessTokenInfoResponseMock->method('getStatusCode')->willReturn(200);

        $clientMock = $this->getMockBuilder(ClientInterface::class)->getMock();
        $clientMock->expects($this->exactly(2))
            ->method('send')
            ->willReturnOnConsecutiveCalls($postResponseMock, $accessTokenInfoResponseMock);

        $this->provider->setHttpClient($clientMock);

        $token = $this->provider->getAccessToken('authorization_code', ['code' => 'mock_authorization_code']);
        $user = $this->provider->getResourceOwner($token);

        $this->assertEquals($accessTokenInfo, $user->toArray());
        $this->assertEquals($accessTokenInfo['userId'], $user->getUserId());
        $this->assertEquals($accessTokenInfo['userId'], $user->getId());
        $this->assertEquals($accessTokenInfo['username'], $user->getUsername());
        $this->assertEquals($accessTokenInfo['accountType'], $user->getAccountType());
        $this->assertEquals($accessTokenInfo['registrationMarketplaceId'], $user->getRegistrationMarketplaceId());
    }

    /**
     * @expectedException \Nitromike502\OAuth2\Client\Provider\Exception\EbayProviderException
     */
    public function testExceptionThrownWhenErrorReceived()
    {
        $status = rand(401,599);
        $postResponseMock = $this->getMockBuilder(ResponseInterface::class)->getMock();
        $postResponseMock->method('getBody')->willReturn('{"error": "error_code","error_description": "A human readable error message"}');
        $postResponseMock->method('getHeader')->willReturn(['content-type' => 'json']);
        $postResponseMock->method('getStatusCode')->willReturn($status);

        $clientMock = $this->getMockBuilder(ClientInterface::class)->getMock();
        $clientMock->expects($this->once())->method('send')->willReturn($postResponseMock);

        $this->provider->setHttpClient($clientMock);

        $this->provider->getAccessToken('authorization_code', ['code' => 'mock_authorization_code']);
    }
}
